cunumeric/vectorize.py

```diff
@@ -0,0 +1,643 @@
+        self._cpu_func: numba.core.ccallback.CFunc
+        self._gpu_func: tuple[Any]
+        self._args: List[Any] = []
+        self._scalar_args: List[Any] = []
+        self._scalar_idxs: List[int] = []
+        self._scalar_names: List[str] = []
+        self._arg_names: List[str] = []
+        self._kwargs: List[Any] = []
+        self._context = runtime.legate_context
+        self._created: bool = False
+        self._func_body: List[str] = []
+
+        if doc is None:
+            self.__doc__ = pyfunc.__doc__
+        else:
+            self.__doc__ = doc
+
+        self._return_names = self._get_return_argumets()
+        self._num_outputs: int = len(self._return_names)
+        self._return_args: List[Any] = []
+        self._output_dtype: Optional[np.dtype[Any]] = None
+        self._cached_dtype: Optional[np.dtype[Any]] = None
+        self._cached_scalar_types: List[Any] = []
+
+        if otypes is not None:
+            if self._num_outputs != len(otypes):
+                raise ValueError(
+                    "number of types in otypes is not consistente"
+                    " with the number of return values defined in pyfunc"
+                )
+            if len(otypes) > 1:
+                for t in otypes:
+                    if t != otypes[0]:
+                        raise NotImplementedError(
+                            "cuNumeric doesn't support variable types"
+                            " in otypes"
+                        )
+                self._output_dtype = np.dtype(otypes[0])
+
+        # FIXME
+        if excluded is not None:
+            raise NotImplementedError(
+                "excluded variables are not supported yet"
+            )
+
+        # FIXME
+        if signature is not None:
+            raise NotImplementedError(
+                "signature variable is not supported yet"
+            )
+
+    def _get_func_body(self, func: Callable[[Any], Any]) -> list[str]:
+        """Using the magic method __doc__, we KNOW the size of the docstring.
+        We then, just subtract this from the total length of the function
+        """
+        lines_to_skip = 0
+        if func.__doc__ is not None and len(func.__doc__.split("\n")) > 0:
+            lines_to_skip = len(func.__doc__.split("\n"))
+
+        lines = inspect.getsourcelines(func)[0]  # type ignore
+
+        return_lines = []
+        for i in range(lines_to_skip + 1, len(lines)):
+            return_lines.append(lines[i].rstrip())
+        return return_lines
+
+    def _get_return_argumets(self) -> list[str]:
+        """
+        Returns the list of names for return arrays/values
+        """
+        self._func_body = self._get_func_body(self._pyfunc)
+        return_names = []
+        for ln in self._func_body:
+            if "return" in ln:
+                ln = ln.replace("return", "")
+                ln = ln.replace(" ", "")
+                return_names = ln.split(",")
+        return return_names
```

Comment:
suggest using tuples instead of lists anywhere explicit mutability is not a requirement 

```suggestion
        return tuple(return_names)
```

Reply:
Do we, in general, want to use tuples instead of lists?

Reply:
In this case, I can make some of the parameters be Tuple and some List. However I combine them together , but at some point I add them.

Reply:
> Do we, in general, want to use tuples instead of lists?

Specifically when functions return sequences, I do strongly prefer to return tuples unless there is a clear reason not to. Even if there is a need to modify the result somewhere down the line, the fact that a tuple was returned will explicitly force any "mutation" to happen to a copy, which is (IMO) the default safe path. 

*Personally*, in the last few years, I do try to prefer tuples *in general*, anywhere there is not a clear motivation or need  to use a mutable type.